### PR TITLE
History of search

### PR DESCRIPTION
**Pull request by sleepergoose**

### CODE REVIEW

File: frontend/src/app/modules/main/search/components/all-albums/all-albums.component.ts [hunk 1]
```diff
@@ -31,8 +35,16 @@ export class AllAlbumsComponent implements OnInit, OnDestroy {
...
   constructor(
     private _searchService: SearchService,
-    private _activatedRoute: ActivatedRoute
-  ) { }
+    private _activatedRoute: ActivatedRoute,
+    private _authService: AuthService,
+    private _searchHistoryService: SearchHistoryService
+  ) {
+    this._authService.getAuthStateObservable()
+      .pipe(filter((state) => !!state))
```
Comment (proskurnicha) [inline frontend/src/app/modules/main/search/components/all-albums/all-albums.component.ts]: Please add takeUntil(this._unsubscribe$) for all subscribe methods example
https://github.com/BinaryStudioAcademy/bsa-2021-perflow/blob/53470250fbc50502dcdc83ec43fd50045595b0de/frontend/src/app/modules/shared/create-edit-playlist/create-edit-playlist/create-edit-playlist.component.ts#L96

File: frontend/src/app/modules/main/search/components/all-albums/all-albums.component.ts [hunk 2]
```diff
@@ -72,4 +84,16 @@ export class AllAlbumsComponent implements OnInit, OnDestroy {
...
   }
+
+  saveToSearchHistory = (album: AlbumForReadDTO) => {
+    const history = {
+      userId: this._userId,
+      albumId: album.id
+    } as WriteSearchHistory;
+
+    this._searchHistoryService.addSearchHistory(history)
+      .subscribe({
```
Comment (proskurnicha) [inline frontend/src/app/modules/main/search/components/all-albums/all-albums.component.ts]: unsubscribe or take(1)

File: frontend/src/app/modules/main/search/components/all-artists/all-artists.component.ts [hunk 3]
```diff
@@ -31,8 +35,16 @@ export class AllArtistsComponent implements OnInit, OnDestroy {
 
   constructor(
     private _searchService: SearchService,
-    private _activatedRoute: ActivatedRoute
-  ) { }
+    private _activatedRoute: ActivatedRoute,
+    private _searchHistoryService: SearchHistoryService,
+    private _authService: AuthService
+  ) {
+    this._authService.getAuthStateObservable()
```
Comment (proskurnicha) [inline frontend/src/app/modules/main/search/components/all-artists/all-artists.component.ts:42]: unsubscribe

File: frontend/src/app/modules/main/search/components/all-artists/all-artists.component.ts [hunk 4]
```diff
@@ -72,4 +84,16 @@ export class AllArtistsComponent implements OnInit, OnDestroy {
...
   }
+
+  saveToSearchHistory = (artist: ArtistReadDTO) => {
+    const history = {
+      userId: this._userId,
+      artistId: artist.id
+    } as WriteSearchHistory;
+
+    this._searchHistoryService.addSearchHistory(history)
+      .subscribe({
```
Comment (proskurnicha) [inline frontend/src/app/modules/main/search/components/all-artists/all-artists.component.ts]: also unsubscribe

File: frontend/src/app/modules/main/cards/playlist-card/playlist-card.component.html [hunk 5]
```diff
@@ -7,6 +7,6 @@
   </div>
   <div class="content">
       <div class="card-name"> {{playlist.name}} </div>
-      <div class="card-desc"> {{playlist.description}} </div>
+      <div class="card-desc"> {{playlist.description === null ? playlist.description : '' }} </div>
```
Comment (proskurnicha) [inline frontend/src/app/modules/main/cards/playlist-card/playlist-card.component.html]: How it works?

File: frontend/src/app/modules/main/search/search.component.ts [hunk 6]
```diff
@@ -41,22 +59,43 @@ export class SearchComponent implements OnInit, OnDestroy {
...
     this._searchTerms$.pipe(
       takeUntil(this._unsubscribe$),
       debounceTime(this.debounceTime),
       distinctUntilChanged(),
-      switchMap((term: string) => this._searchService.getFoundData(term))
+      switchMap((term: string) => {
+        if (term) {
+          return this._searchService.getFoundData(term);
+        }
+        return new Array<FoundData>();
```
Comment (proskurnicha) [inline frontend/src/app/modules/main/search/search.component.ts]: switchMap should return observable
`
return of(new Array<FoundData>())
`

File: frontend/src/app/modules/main/search/components/search-artists-result/search-artists-result.component.ts [hunk 7]
```diff
@@ -9,4 +14,34 @@ import { ArtistReadDTO } from 'src/app/models/user/ArtistReadDTO';
...
+  saveToSearchHistory = (artist: ArtistReadDTO) => {
+    const history = {
+      userId: this._userId,
+      artistId: artist.id
+    } as WriteSearchHistory;
+
+    this._searchHistoryService.addSearchHistory(history)
+      .pipe(takeUntil(this._unsubscribe$))
+      .subscribe({
+        next: () => {
```
Comment (proskurnicha) [inline frontend/src/app/modules/main/search/components/search-artists-result/search-artists-result.component.ts]: Why do you call into next this._unsubscribe$.next()
If you need only one value write use take \ first operators
`
this._searchHistoryService.addSearchHistory(history)
      .pipe(take(1)).subscribe()
`

File: frontend/src/app/modules/main/search/components/search-albums-result/search-albums-result.component.ts [hunk 8]
```diff
@@ -9,4 +14,34 @@ import { AlbumForReadDTO } from 'src/app/models/album/albumForReadDTO';
...
+  }
+
+  saveToSearchHistory = (album: AlbumForReadDTO) => {
+    const history = {
+      userId: this._userId,
+      albumId: album.id
+    } as WriteSearchHistory;
+
+    this._searchHistoryService.addSearchHistory(history)
+      .pipe(takeUntil(this._unsubscribe$))
```
Comment (proskurnicha) [inline frontend/src/app/modules/main/search/components/search-albums-result/search-albums-result.component.ts]: same

File: frontend/src/app/modules/main/cards/playlist-card/playlist-card.component.html (hunk 5 above)
Comment (sleepergoose) [reply frontend/src/app/modules/main/cards/playlist-card/playlist-card.component.html]: There was such a problem: when playlist.description=null (empty in database) then 'null' is shown on the playlist card. To resolve this problem I tried playlist.description ?? '' but this didn't work. I'll try to hide description block by *ngIf

Comment (proskurnicha) [reply frontend/src/app/modules/main/cards/playlist-card/playlist-card.component.html]: you could try this playlist?.description

**Review by proskurnicha — APPROVED** [review]

File: frontend/src/app/modules/main/search/search.component.ts [hunk 9]
```diff
@@ -76,7 +77,9 @@ export class SearchComponent implements OnInit, OnDestroy {
       .subscribe({
         next: (data) => {
           this.searchHistory = data;
-          this.isSearchHistoryShown = true;
+          if (data.length) {
+            this.isSearchHistoryShown = true;
+          }
```
Comment (proskurnicha) [inline frontend/src/app/modules/main/search/search.component.ts]: You could use
  this.isSearchHistoryShown = !!data.length;

**Review by proskurnicha — APPROVED** [review]
Looks good
Did you try playlist?.description instead of *ngif?

**Comment by sleepergoose** [comment]
Yes, I tried playlist?.description and the others, but they didn't work

**Pull request merged**